presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1583,11 +1587,45 @@ public Expression rewriteIdentifier(Identifier reference, Void context, Expressi
             }
         }
 
+        private void checkGroupingSetsCount(GroupBy node)
+        {
+            // If groupBy is distinct then crossProduct will be overestimated if there are duplicate grouping sets.
+            int crossProduct = 1;
+            for (GroupingElement element : node.getGroupingElements()) {
+                try {
+                    int product;
+                    if (element instanceof Cube) {
+                        int exponent = ((Cube) element).getColumns().size();
+                        if (exponent > 30) {
+                            throw new ArithmeticException();
+                        }
+                        product = 1 << exponent;
+                    }
+                    else if (element instanceof Rollup) {
+                        product = ((Rollup) element).getColumns().size() + 1;
+                    }
+                    else {
+                        product = element.enumerateGroupingSets().size();
+                    }
+                    crossProduct = Math.multiplyExact(crossProduct, product);
+                }
+                catch (ArithmeticException e) {
+                    throw new SemanticException(INVALID_PROCEDURE_ARGUMENTS, node,
```

Comment:
This error is not appropriate. Add a TOO_MANY_GROUPING_SETS to SemanticErrorCode.